test/profiler/test_profiler.py

```diff
@@ -549,6 +553,161 @@ def extract(pattern: str):
 
         torch._C._set_graph_executor_optimize(prev_opt)
 
+    @parametrize(
+        "name,thread_spec",
+        {
+            "basic": ((False, False),),
+            "multiple_preexisting": ((False, False), ) * 2,
+            "open_in_scope": ((True, False),),
+            "close_in_scope": ((False, True),),
+            "complex": (
+                # Large number of background threads
+                (False, False),
+                (False, False),
+                (False, False),
+                (False, False),
+
+                # some of which finish during profiling
+                (False, True),
+                (False, True),
+
+                # And the profiled section is also multithreaded
+                (True, False),
+                (True, True),
+
+            ),
+        }.items(),
+        name_fn=lambda name, thread_spec: name
+    )
+    @parametrize("work_in_main_thread", [True, False])
+    def test_source_multithreaded(self, name, thread_spec, work_in_main_thread):
```

Comment:
Hi, @robieta 
May i know if the case `test_source_multithreaded` has the internal connection with the other test cases in the same test class `TestProfiler`? We find the case `profiler/test_profiler.py::TestProfilerMultithreaded::test_source_multithreaded_basic_work_in_main_thread_True` will failed if moving the test case `test_source_multithreaded` outside of the test class `TestProfiler`.
Here is the test PR: https://github.com/pytorch/pytorch/pull/135971
Could you help take a look?
Thank you.

Reply:
Here is the error log i got in CI:
```
2024-09-16T02:54:44.0637567Z   File "/var/lib/jenkins/workspace/test/profiler/test_profiler.py", line 228, in test_source_multithreaded
2024-09-16T02:54:44.0678730Z #22 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0679126Z #23 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0679599Z #24 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0680087Z #25 method_vectorcall from :0
2024-09-16T02:54:44.0680454Z #26 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0680838Z #27 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0681324Z #28 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0681723Z #29 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0682132Z #30 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0682585Z #31 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0683028Z #32 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0683413Z #33 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0683837Z #34 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0684288Z #35 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0684732Z #36 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0685147Z #37 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0685635Z #38 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0686103Z #39 method_vectorcall from :0
2024-09-16T02:54:44.0686476Z #40 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0686889Z #41 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0687274Z #42 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0687689Z #43 PyObject_Vectorcall from ??:0
2024-09-16T02:54:44.0688082Z #44 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0688550Z #45 _PyObject_VectorcallTstate.lto_priv.4.cold from :0
2024-09-16T02:54:44.0689028Z #46 method_vectorcall from :0
2024-09-16T02:54:44.0689398Z #47 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0689801Z #48 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0690226Z #49 eval_custom_code from eval_frame.c:0
2024-09-16T02:54:44.0690677Z #50 custom_eval_frame_shim from eval_frame.c:0
2024-09-16T02:54:44.0691115Z #51 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0691530Z #52 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0691995Z #53 _PyObject_VectorcallTstate.lto_priv.4 from :0
2024-09-16T02:54:44.0692440Z #54 method_vectorcall from :0
2024-09-16T02:54:44.0692810Z #55 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0693287Z #56 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0693765Z #57 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0694259Z #58 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0694651Z #59 slot_tp_call from :0
2024-09-16T02:54:44.0695009Z #60 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0695463Z #61 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0695944Z #62 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0696422Z #63 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0696893Z #64 slot_tp_call from :0
2024-09-16T02:54:44.0697242Z #65 _PyObject_Call.localalias from :0
2024-09-16T02:54:44.0697647Z #66 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0698130Z #67 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0698630Z #68 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0699007Z #69 slot_tp_call from :0
2024-09-16T02:54:44.0699377Z #70 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0699831Z #71 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0700307Z #72 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0700793Z #73 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0701179Z #74 slot_tp_call from :0
2024-09-16T02:54:44.0701534Z #75 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0701984Z #76 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0702465Z #77 _PyObject_FastCallDictTstate.localalias from :0
2024-09-16T02:54:44.0702941Z #78 _PyObject_Call_Prepend from ??:0
2024-09-16T02:54:44.0703329Z #79 slot_tp_call from :0
2024-09-16T02:54:44.0703699Z #80 _PyObject_MakeTpCall.localalias from :0
2024-09-16T02:54:44.0714578Z #81 _PyEval_EvalFrameDefault from ??:0
2024-09-16T02:54:44.0715148Z #82 PyEval_EvalCode from ??:0
2024-09-16T02:54:44.0715521Z #83 run_eval_code_obj from :0
2024-09-16T02:54:44.0715867Z #84 run_mod from :0
2024-09-16T02:54:44.0716155Z #85 pyrun_file from :0
2024-09-16T02:54:44.0716549Z #86 _PyRun_SimpleFileObject.localalias from :0
2024-09-16T02:54:44.0717050Z #87 _PyRun_AnyFileObject.localalias from :0
2024-09-16T02:54:44.0717624Z #88 Py_RunMain.localalias from :0
2024-09-16T02:54:44.0718014Z #89 Py_BytesMain from ??:0
2024-09-16T02:54:44.0718364Z #90 __libc_start_main from ??:0
2024-09-16T02:54:44.0718712Z #91 _start from ??:0
```